lib/auth/keystore/manager.go

```diff
@@ -305,6 +305,9 @@ func publicKeyFromSSHAuthorizedKey(sshAuthorizedKey []byte) (crypto.PublicKey, e
 // toRSASHA512Signer forces an ssh.MultiAlgorithmSigner into using
 // "rsa-sha2-sha512" (instead of its SHA256 default).
 func toRSASHA512Signer(signer ssh.Signer) ssh.Signer {
+	if signer.PublicKey().Type() != ssh.KeyAlgoRSA {
```

Comment:
Aren't we doing the same check in the caller of `toRSASHA512Signer`?

Reply:
we do in master, but not in branch/v16, which is the base branch of this PR. Current v16 is broken for all non-RSA keys (but we only ever generate RSA keys so its mostly fine). This is a much simplified version of https://github.com/gravitational/teleport/pull/45887 which updates `toRSASHA512Signer` to a new `sshSignerFromCryptoSigner` which specifically selects signature algorithms for all key types and sizes